x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInferModelAction.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.action;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.ml.action.InferModelAction;
+import org.elasticsearch.xpack.core.ml.inference.results.InferenceResults;
+import org.elasticsearch.xpack.ml.inference.loadingservice.Model;
+import org.elasticsearch.xpack.ml.inference.loadingservice.ModelLoadingService;
+import org.elasticsearch.xpack.ml.utils.TypedChainTaskExecutor;
+
+
+public class TransportInferModelAction extends HandledTransportAction<InferModelAction.Request, InferModelAction.Response> {
+
+    private final ModelLoadingService modelLoadingService;
+    private final Client client;
+
+    @Inject
+    public TransportInferModelAction(TransportService transportService,
+                                     ActionFilters actionFilters,
+                                     ModelLoadingService modelLoadingService,
+                                     Client client) {
+        super(InferModelAction.NAME, transportService, actionFilters, InferModelAction.Request::new);
+        this.modelLoadingService = modelLoadingService;
+        this.client = client;
+    }
+
+    @Override
+    protected void doExecute(Task task, InferModelAction.Request request, ActionListener<InferModelAction.Response> listener) {
+
+        ActionListener<Model> getModelListener = ActionListener.wrap(
+            model -> {
+                TypedChainTaskExecutor<InferenceResults<?>> typedChainTaskExecutor =
+                    new TypedChainTaskExecutor<>(client.threadPool().executor(ThreadPool.Names.SAME),
+                    // run through all tasks
+                    r -> true,
+                    // Always fail immediately and return an error
+                    ex -> true);
+                if (request.getTopClasses() != null) {
+                    request.getObjectsToInfer().forEach(stringObjectMap ->
```

Comment:
So we don't do `infer` when topClasses is set? I was expecting that top classes would be additional info.

Reply:
`infer` is a convenience method that (right now) either:

* Returns the top class
* Returns the regression value

`topClasses` should only be set on classification models. If it is requested against a regression model, we throw an exception.